File: modules/kasane-gd/src/mesh_data.cpp
// SPDX-License-Identifier: MIT
#include "mesh_data.hpp"
#include "document_bridge.hpp"
#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/core/object.hpp>
#include <godot_cpp/classes/os.hpp>
using namespace godot;

namespace kasane_gd {
void KasaneMeshData::_bind_methods() {
    ClassDB::bind_method(D_METHOD("is_valid"), &KasaneMeshData::is_valid);
    ClassDB::bind_method(D_METHOD("get_id"), &KasaneMeshData::get_id);
    ClassDB::bind_method(D_METHOD("snapshot"), &KasaneMeshData::snapshot);
    ClassDB::bind_method(D_METHOD("get_name"), &KasaneMeshData::get_name);
    ClassDB::bind_method(D_METHOD("set_name", "value"), &KasaneMeshData::set_name);
    ClassDB::bind_method(D_METHOD("get_positions"), &KasaneMeshData::get_positions);
    ClassDB::bind_method(D_METHOD("set_positions", "value"), &KasaneMeshData::set_positions);
    ClassDB::bind_method(D_METHOD("get_vertex_ids"), &KasaneMeshData::get_vertex_ids);
    ClassDB::bind_method(D_METHOD("set_vertex_positions", "ids", "positions"),
                         &KasaneMeshData::set_vertex_positions);
    ClassDB::bind_method(D_METHOD("replace_geometry", "ids", "positions", "uvs", "triangles"),
                         &KasaneMeshData::replace_geometry);
    ADD_PROPERTY(PropertyInfo(Variant::STRING, "id"), "", "get_id");
    ADD_PROPERTY(PropertyInfo(Variant::STRING, "name"), "set_name", "get_name");
    ADD_PROPERTY(PropertyInfo(Variant::PACKED_VECTOR2_ARRAY, "positions"), "set_positions", "get_positions");
    ADD_PROPERTY(PropertyInfo(Variant::PACKED_INT64_ARRAY, "vertex_ids"), "", "get_vertex_ids");
}

void KasaneMeshData::attach(uint64_t owner_id, uint64_t generation, const String &id) {
    owner_ = owner_id;
    generation_ = generation;
    id_ = id;
}

KasaneDocumentBridge *KasaneMeshData::owner() const {
    if (OS::get_singleton()->get_thread_caller_id() != OS::get_singleton()->get_main_thread_id())
        return nullptr;
    auto *bridge = Object::cast_to<KasaneDocumentBridge>(ObjectDB::get_instance(owner_));
    return bridge && bridge->generation() == generation_ ? bridge : nullptr;
}

bool KasaneMeshData::is_valid() const {
    auto *bridge = owner();
    return bridge && bool(bridge->get_mesh_snapshot(id_)["ok"]);
}

Dictionary KasaneMeshData::snapshot() const {
    auto *bridge = owner();
    return bridge ? bridge->get_mesh_snapshot(id_)
                  : error("STALE_HANDLE", "The owning document was closed or replaced.");
}

String KasaneMeshData::get_name() const {
    return snapshot().get("name", String());
}

void KasaneMeshData::set_name(const String &value) {
    auto *bridge = owner();
    ERR_FAIL_NULL_MSG(bridge, "The owning document was closed or replaced.");
    Dictionary edit = bridge->rename_mesh(id_, value);
    ERR_FAIL_COND_MSG(!bool(edit["ok"]), String(edit["message"]));
}

PackedVector2Array KasaneMeshData::get_positions() const {
    return snapshot().get("base_positions", PackedVector2Array());
}

PackedInt64Array KasaneMeshData::get_vertex_ids() const {
    return snapshot().get("vertex_ids", PackedInt64Array());
}

void KasaneMeshData::set_positions(const PackedVector2Array &value) {
    Dictionary edit = set_vertex_positions(get_vertex_ids(), value);
    ERR_FAIL_COND_MSG(!bool(edit["ok"]), String(edit["message"]));
}

Dictionary KasaneMeshData::set_vertex_positions(const PackedInt64Array &vertices,
                                                const PackedVector2Array &positions) {
    auto *bridge = owner();
    return bridge ? bridge->set_vertex_positions(id_, vertices, positions)
                  : error("STALE_HANDLE", "The owning document was closed or replaced.");
}

Dictionary KasaneMeshData::replace_geometry(const PackedInt64Array &vertices,
                                            const PackedVector2Array &positions,
                                            const PackedVector2Array &uvs,
                                            const PackedInt64Array &triangles) {
    auto *bridge = owner();
    if (!bridge)
        return error("STALE_HANDLE", "The owning document was closed or replaced.");
    auto data = snapshot();
    if (!bool(data["ok"]))
        return data;
    data["vertex_ids"] = vertices;
    data["base_positions"] = positions;
    data["uvs"] = uvs;
    data["triangles"] = triangles;
    return bridge->replace_mesh(data);
}
} // namespace kasane_gd

File: modules/kasane-godot/src/mesh_data.rs
use godot::obj::InstanceId;
use godot::prelude::*;

use crate::conversions::{error_dict, is_main_thread, Dictionary};
use crate::document_bridge::KasaneDocumentBridge;

#[derive(GodotClass)]
#[class(init, base=RefCounted)]
pub struct KasaneMeshData {
    base: Base<RefCounted>,
    owner_id: u64,
    generation: u64,
    id: GString,
}

#[godot_api]
impl KasaneMeshData {
    pub fn attach(&mut self, owner: u64, generation: u64, id: GString) {
        self.owner_id = owner;
        self.generation = generation;
        self.id = id;
    }

    fn owner(&self) -> Option<Gd<KasaneDocumentBridge>> {
        if !is_main_thread() {
            return None;
        }
        let instance_id = InstanceId::try_from_i64(self.owner_id as i64)?;
        let bridge = Gd::<KasaneDocumentBridge>::try_from_instance_id(instance_id).ok()?;
        if bridge.bind().generation() == self.generation {
            Some(bridge)
        } else {
            None
        }
    }

    #[func]
    pub fn is_valid(&self) -> bool {
        let Some(bridge) = self.owner() else {
            return false;
        };
        let snap = bridge.bind().get_mesh_snapshot(self.id.clone());
        snap.get("ok")
            .and_then(|v| v.try_to::<bool>().ok())
            .unwrap_or(false)
    }

    #[func]
    pub fn get_id(&self) -> GString {
        self.id.clone()
    }

    #[func]
    pub fn snapshot(&self) -> Dictionary {
        let Some(bridge) = self.owner() else {
            return error_dict(
                "STALE_HANDLE",
                "The owning document was closed or replaced.",
            );
        };
        let res = bridge.bind().get_mesh_snapshot(self.id.clone());
        res
    }

    #[func]
    pub fn get_name(&self) -> GString {
        self.snapshot()
            .get("name")
            .and_then(|v| v.try_to::<GString>().ok())
            .unwrap_or_default()
    }

    #[func]
    pub fn set_name(&mut self, value: GString) {
        let Some(mut bridge) = self.owner() else {
            godot_error!("The owning document was closed or replaced.");
            return;
        };
        let edit = bridge.bind_mut().rename_mesh(self.id.clone(), value);
        let ok = edit
            .get("ok")
            .and_then(|v| v.try_to::<bool>().ok())
            .unwrap_or(false);
        if !ok {
            let msg = edit
                .get("message")
                .and_then(|v| v.try_to::<GString>().ok())
                .unwrap_or_default();
            godot_error!("{}", msg);
        }
    }

    #[func]
    pub fn get_positions(&self) -> PackedVector2Array {
        self.snapshot()
            .get("base_positions")
            .and_then(|v| v.try_to::<PackedVector2Array>().ok())
            .unwrap_or_default()
    }

    #[func]
    pub fn set_positions(&mut self, value: PackedVector2Array) {
        let edit = self.set_vertex_positions(self.get_vertex_ids(), value);
        let ok = edit
            .get("ok")
            .and_then(|v| v.try_to::<bool>().ok())
            .unwrap_or(false);
        if !ok {
            let msg = edit
                .get("message")
                .and_then(|v| v.try_to::<GString>().ok())
                .unwrap_or_default();
            godot_error!("{}", msg);
        }
    }

    #[func]
    pub fn get_vertex_ids(&self) -> PackedInt64Array {
        self.snapshot()
            .get("vertex_ids")
            .and_then(|v| v.try_to::<PackedInt64Array>().ok())
            .unwrap_or_default()
    }

    #[func]
    pub fn set_vertex_positions(
        &mut self,
        ids: PackedInt64Array,
        positions: PackedVector2Array,
    ) -> Dictionary {
        let Some(mut bridge) = self.owner() else {
            return error_dict(
                "STALE_HANDLE",
                "The owning document was closed or replaced.",
            );
        };
        let res = bridge
            .bind_mut()
            .set_vertex_positions(self.id.clone(), ids, positions);
        res
    }

    #[func]
    pub fn replace_geometry(
        &mut self,
        ids: PackedInt64Array,
        positions: PackedVector2Array,
        uvs: PackedVector2Array,
        triangles: PackedInt64Array,
    ) -> Dictionary {
        let Some(mut bridge) = self.owner() else {
            return error_dict(
                "STALE_HANDLE",
                "The owning document was closed or replaced.",
            );
        };
        let mut data = self.snapshot();
        let ok = data
            .get("ok")
            .and_then(|v| v.try_to::<bool>().ok())
            .unwrap_or(false);
        if !ok {
            return data;
        }
        data.set("vertex_ids", &ids);
        data.set("base_positions", &positions);
        data.set("uvs", &uvs);
        data.set("triangles", &triangles);
        let res = bridge.bind_mut().replace_mesh(data);
        res
    }
}
